src/backup.py

```diff
@@ -128,3 +161,255 @@ def can_use_bucket(self) -> bool:
+
+    logger.info(
+        "Calculating size of paths. Command: %s. stdout: %s. stderr: %s", command, stdout, stderr
+    )
+
+    try:
+        return int(stdout)
+    except ValueError as exc:
+        raise BackupError("Cannot calculate size of paths. Wrong stdout.") from exc
+
+
+def _build_backup_command(
+    s3_parameters: S3Parameters,
+    backup_id: str,
+    backup_paths: Iterable[str],
+    passphrase_file: str,
+    expected_size: int = int(1e10),
+) -> List[str]:
+    """Build the command to execute the backup.
+
+    Args:
+        s3_parameters: S3 parameters.
+        backup_id: The name of the object to back up.
+        backup_paths: List of paths to back up.
+        passphrase_file: Passphrase to use to encrypt the backup file.
+        expected_size: expected size of the backup, so AWS S3 Client can calculate
+            a reasonable size for the upload parts.
+
+    Returns:
+        The backup command to execute.
+    """
+    bash_strict_command = "set -euxo pipefail; "
+    paths = _paths_to_args(backup_paths)
+    tar_command = f"tar -c {paths}"
+    gpg_command = f"gpg --batch --no-symkey-cache --passphrase-file '{passphrase_file}'"
+    gpg_command += " --symmetric"
+    aws_command = f"{AWS_COMMAND} s3 cp --expected-size={expected_size} - "
+
+    s3_url = _s3_path(
+        prefix=s3_parameters.path, object_name=backup_id, bucket=s3_parameters.bucket
+    )
+    aws_command += f"'{s3_url}'"
+    full_command = bash_strict_command + " | ".join((tar_command, gpg_command, aws_command))
+    return [BASH_COMMAND, "-c", full_command]
+
+
+def _get_environment(s3_parameters: S3Parameters) -> Dict[str, str]:
+    """Get the environment variables for backup that configure aws S3 cli.
+
+    Args:
+        s3_parameters: S3 parameters.
+
+    Returns:
+        A dictionary with aws s3 configuration variables.
+    """
+    environment = {
+        "AWS_ACCESS_KEY_ID": s3_parameters.access_key,
+        "AWS_SECRET_ACCESS_KEY": s3_parameters.secret_key,
+    }
+    if s3_parameters.endpoint:
+        environment["AWS_ENDPOINT_URL"] = s3_parameters.endpoint
+    if s3_parameters.region:
+        environment["AWS_DEFAULT_REGION"] = s3_parameters.region
+    return environment
+
+
+def _s3_path(prefix: str, object_name: Optional[str] = None, bucket: Optional[str] = None) -> str:
+    """Create a S3 paths compatible with S3 and Minio for backup purposes.
+
+    Args:
+        prefix: prefix for the path.
+        object_name: final part of the name of the object.
+        bucket: name of the bucket.
+
+    Returns:
+       a S3 path, only including the protocol if the bucket is included.
+    """
+    # Removing the root slash in path makes this code compatible with
+    # MinIO. MinIO looks like it does not have a slash as the first element,
+    # and if the file was uploaded with one, it gets removed.
+    path = prefix.strip("/")
+    if object_name:
+        path = str(pathlib.Path(f"{path}/{object_name}"))
```

Comment:
Why instantiating a path to convert it again to a string?

Reply:
I was thinking in only having valid paths. That is it, cpnverting "path//subpath" to "path/subpath". It is because of this: https://github.com/minio/minio/issues/5958

Reply:
Oh! Makes sense. Would you mind adding a comment explaining the reasoning?